File: sections/72_master_flow_convex_mixed/problems/a_exam_room_bounds/solution.py
import collections
import sys


class Dinic:
    def __init__(self, n):
        self.g = [[] for _ in range(n)]

    def add_edge(self, u, v, cap):
        self.g[u].append([v, cap, len(self.g[v])])
        self.g[v].append([u, 0, len(self.g[u]) - 1])

    def flow(self, s, t):
        answer = 0
        while True:
            level = [-1] * len(self.g)
            level[s] = 0
            q = collections.deque([s])
            while q:
                u = q.popleft()
                for v, cap, _ in self.g[u]:
                    if cap and level[v] < 0:
                        level[v] = level[u] + 1
                        q.append(v)
            if level[t] < 0:
                return answer
            it = [0] * len(self.g)

            def dfs(u, pushed):
                if u == t:
                    return pushed
                while it[u] < len(self.g[u]):
                    e = self.g[u][it[u]]
                    v, cap, rev = e
                    if cap and level[v] == level[u] + 1:
                        take = dfs(v, min(pushed, cap))
                        if take:
                            e[1] -= take
                            self.g[v][rev][1] += take
                            return take
                    it[u] += 1
                return 0

            while pushed := dfs(s, 10**30):
                answer += pushed


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    rows, cols, m = data[:3]
    at = 3
    demands = data[at : at + rows]
    at += rows
    bounds = []
    for _ in range(cols):
        bounds.append(tuple(data[at : at + 2]))
        at += 2
    n = rows + cols + 2
    source, sink, ss, tt = rows + cols, rows + cols + 1, n, n + 1
    dinic = Dinic(n + 2)
    balance = [0] * n

    def bounded(u, v, low, high):
        dinic.add_edge(u, v, high - low)
        balance[u] -= low
        balance[v] += low

    total = sum(demands)
    for r, demand in enumerate(demands):
        bounded(source, r, demand, demand)
    for c, (low, high) in enumerate(bounds):
        bounded(rows + c, sink, low, high)
    for _ in range(m):
        r, c = data[at : at + 2]
        at += 2
        bounded(r - 1, rows + c - 1, 0, 1)
    bounded(sink, source, total, total)
    need = 0
    for v, value in enumerate(balance):
        if value > 0:
            dinic.add_edge(ss, v, value)
            need += value
        elif value < 0:
            dinic.add_edge(v, tt, -value)
    print("YES" if dinic.flow(ss, tt) == need else "NO")


if __name__ == "__main__":
    main()

File: sections/72_master_flow_convex_mixed/problems/a_exam_room_bounds/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

struct Dinic {
  struct Edge {
    int to, rev;
    int64 cap;
  };
  vector<vector<Edge>> g;
  vector<int> level, it;
  Dinic(int n) : g(n), level(n), it(n) {}
  void add_edge(int u, int v, int64 cap) {
    Edge a{v, (int)g[v].size(), cap};
    Edge b{u, (int)g[u].size(), 0};
    g[u].push_back(a);
    g[v].push_back(b);
  }
  bool bfs(int s, int t) {
    fill(level.begin(), level.end(), -1);
    queue<int> q;
    level[s] = 0;
    q.push(s);
    while (!q.empty()) {
      int u = q.front();
      q.pop();
      for (auto &e : g[u])
        if (e.cap && level[e.to] < 0)
          level[e.to] = level[u] + 1, q.push(e.to);
    }
    return level[t] >= 0;
  }
  int64 dfs(int u, int t, int64 pushed) {
    if (u == t)
      return pushed;
    for (int &i = it[u]; i < (int)g[u].size(); ++i) {
      Edge &e = g[u][i];
      if (!e.cap || level[e.to] != level[u] + 1)
        continue;
      int64 take = dfs(e.to, t, min(pushed, e.cap));
      if (take) {
        e.cap -= take;
        g[e.to][e.rev].cap += take;
        return take;
      }
    }
    return 0;
  }
  int64 flow(int s, int t) {
    int64 ans = 0, pushed;
    while (bfs(s, t)) {
      fill(it.begin(), it.end(), 0);
      while ((pushed = dfs(s, t, (1LL << 62))))
        ans += pushed;
    }
    return ans;
  }
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int rows, cols, m;
  if (!(cin >> rows >> cols >> m))
    return 0;
  int n = rows + cols + 2;
  int source = rows + cols, sink = source + 1, ss = n, tt = n + 1;
  Dinic dinic(n + 2);
  vector<int64> balance(n);
  int64 total = 0;
  auto bounded = [&](int u, int v, int64 low, int64 high) {
    dinic.add_edge(u, v, high - low);
    balance[u] -= low;
    balance[v] += low;
  };
  for (int i = 0; i < rows; ++i) {
    int64 demand;
    cin >> demand;
    bounded(source, i, demand, demand);
    total += demand;
  }
  for (int j = 0; j < cols; ++j) {
    int64 low, high;
    cin >> low >> high;
    bounded(rows + j, sink, low, high);
  }
  for (int i = 0, r, c; i < m; ++i) {
    cin >> r >> c;
    bounded(--r, rows + --c, 0, 1);
  }
  bounded(sink, source, total, total);
  int64 need = 0;
  for (int v = 0; v < n; ++v) {
    if (balance[v] > 0)
      dinic.add_edge(ss, v, balance[v]), need += balance[v];
    if (balance[v] < 0)
      dinic.add_edge(v, tt, -balance[v]);
  }
  cout << (dinic.flow(ss, tt) == need ? "YES\n" : "NO\n");
}
